### PR TITLE
Add draft post "Operator Constraints"

### PR DESCRIPTION
**Pull request by Merovius**
A rendered version can be viewed here: https://website-review.merovius.de/pull/constraints/posts/draft-operator-constraints/

I'm mainly looking for general understandability and structural feedback, but any comment is welcome.

### CODE REVIEW

File: content/posts/draft-operator-constraints.markdown [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+## `constraints.Ordered`
+
+Go 1.18 has a mechanism to constraint a type parameter to all types which have
+the `<` operator defined on them. The types which have this operator are
+exactly all types whose underlying type is `string` or one of the predeclared
+integer and float types. So we can write a type set expressing that:
+
+```go
+type Integer interface {
+  ~int | ~int8 | ~int16 | ~int32 | ~int64 | ~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uintptr
```
Comment (johanbrandhorst) [inline content/posts/draft-operator-constraints.markdown:30]: The syntax highlighting on your page makes this look pretty ugly 😬.

File: content/posts/draft-operator-constraints.markdown [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+about four in particular here:
+
+1. `constraints.Ordered`
+2. A method constraint
+3. Taking a comparison function
+4. Comparator types
+
+I believe the fourth is a novel idea. At least I don't think I have seen it
+come up in a discussion about Go so far. Though I think I have seen it used in
+other languages.
```
Comment (johanbrandhorst) [inline content/posts/draft-operator-constraints.markdown]: This paragraph feels a bit out of place here. Maybe move it to the comparator types section?

File: content/posts/draft-operator-constraints.markdown [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+## `constraints.Ordered`
+
+Go 1.18 has a mechanism to constraint a type parameter to all types which have
```
Comment (johanbrandhorst) [inline content/posts/draft-operator-constraints.markdown]:
```suggestion
Go 1.18 has a mechanism to constrain a type parameter to all types which have
```

File: content/posts/draft-operator-constraints.markdown [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+}
+```
+
+The disadvantage of this is that it requires some boiler plate on part of your
+user. Using a custom sorting order always requires defining a type with a method.
+
+They can't use your code with predeclared types like `int` or `string` but
+always have to wrap it into a new type.
+
+Likewise, if a type already has a natural comparison method but it is not
```
Comment (johanbrandhorst) [inline content/posts/draft-operator-constraints.markdown]: This comma feels out of place when the sentence is so short. Either remove the comma or merge the first and second sentence.
```suggestion
Likewise if a type already has a natural comparison method but it is not
```

File: content/posts/draft-operator-constraints.markdown [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+to the comparison used. It should be “a set of `string`” not “a set of
+`MyCustomlyOrderedString`”. This reflects the fact that with the method
+constraint, we have to convert back-and-forth when putting things into the
+container or taking it out again.
+
+The main *disadvantage* of this approach is that it means you can not have
+useful zero values. Your `SearchTree` type needs the `Less` field to be
+populated to work. So its zero value can not be used to represent an empty set.
+
+You can not even lazily initialize it (which is a common trick to make types
```
Comment (johanbrandhorst) [inline content/posts/draft-operator-constraints.markdown]:
```suggestion
You cannot even lazily initialize it (which is a common trick to make types
```
Or `can't`

File: content/posts/draft-operator-constraints.markdown [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+with the above ones:
+
+```go
+type LessOperator[T constraints.Ordered] struct{}
+
+func (LessOperator[T]) Less(a, b T) bool {
+  return a < b
+}
+
+type LessMethod[T Lesser[T]] struct{}
```
Comment (johanbrandhorst) [inline content/posts/draft-operator-constraints.markdown:269]: Would it be worth defining the `Lesser` interface too?

File: content/posts/draft-operator-constraints.markdown [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+  return a.Less(b)
+}
+
+type Reversed[C Comparator[T], T any] struct{}
+
+func (Reversed[C, T]) Less(a, b T) bool {
+  var c C
+  return c.Less(b, a)
+}
+```
```
Comment (johanbrandhorst) [inline content/posts/draft-operator-constraints.markdown:281]: Is it possible to define a helper that takes a less function literal?

File: content/posts/draft-operator-constraints.markdown [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+and idea behind the `Comparator` type is likely to trip up your users when
+reading the documentation.
+
+## Summary
+
+We are left with these trade-offs:
+
+|| `constraints.Ordered` | `Lesser[T]` | `func(T,T) bool` | `Comparator[T]` |
+|--------------------------|----|--------|----|----|
+| Predeclared types        | 👍 | 👎     | 👎 | 👎 |
```
Comment (johanbrandhorst) [inline content/posts/draft-operator-constraints.markdown:312]: The thumbs up and down are hard to differentiate on your page, maybe use ✅ and ❌?

**Review by johanbrandhorst — COMMENTED** [review]
This was super interesting! I too agree that this is a novel solution, but also agree that it's a bit rough to use.

File: content/posts/draft-operator-constraints.markdown (hunk 1 above)
Comment (Merovius) [reply content/posts/draft-operator-constraints.markdown:30]: [Yeah, I know](https://twitter.com/TheMerovius/status/1526285257942638594). I should use a custom syntax highlighting color scheme that is less garish.

File: content/posts/draft-operator-constraints.markdown (hunk 2 above)
Comment (Merovius) [reply content/posts/draft-operator-constraints.markdown]: The rationale of that sentence is to entice people to continue reading til the end - that is, I don't want them to open the post and think "okay, same old" and close it again. At the same time, that section must be at the end, so it is clear what it's trying to "fix".

That is, if this reads strange, I'd just remove it. But I'd like some way to make sure the reader gets to that section. Any ideas?

File: content/posts/draft-operator-constraints.markdown (hunk 6 above)
Comment (Merovius) [reply content/posts/draft-operator-constraints.markdown:269]: I was trying to implicitly refer back to the one used in the second section. Maybe that's not clear, but I think I'm fine with that.

File: content/posts/draft-operator-constraints.markdown (hunk 7 above)
Comment (Merovius) [reply content/posts/draft-operator-constraints.markdown:281]: Unfortunately no. Its zero value would not be callable. I will add a sentence to be explicit.

File: content/posts/draft-operator-constraints.markdown (hunk 8 above)
Comment (Merovius) [reply content/posts/draft-operator-constraints.markdown:312]: The checkmark and cross read as "yes or no" to me, while the thumbs read as "good or bad". That is, using the thumbs saves the extra step of having to decode which is "better". For example, for "Predeclared types" a "yes" is better, while for "type boilerplate" a "no" is better.

Alternatively, I would have to find phrasings for every line which mean "yes is always better", which also have to be short enough. I failed at that, myself :)

**Review by mdlayher — APPROVED** [review]
Looks great! The Comparator approach is pretty neat.

File: content/posts/draft-operator-constraints.markdown (hunk 1 above)
Comment (Merovius) [reply content/posts/draft-operator-constraints.markdown:30]: Changed the syntax highlighting to be differently ugly.

**Comment by Merovius** [comment]
Published now, closing this.

File: content/posts/draft-operator-constraints.markdown (hunk 2 above)
Comment (johanbrandhorst) [reply content/posts/draft-operator-constraints.markdown]: I thought about this again after reading the post and I changed my mind - it is nice to whet the readers appetite.

**Pull request closed without merging**